# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: js/src/frontend/TryEmitter.cpp
# Commit: eeaeb40f8b86
# Full Hash: eeaeb40f8b869121eff65b1c2d11af423c9a9ecc
# Author: Iain Ireland <iireland@mozilla.com>
# Date: 2022-03-09 09:44:44
# Regressor Bug: 885514
# File Overlap Count: 1
# Description:
#   Bug 885514: Simplify JSOp::Retsub r=jandem
#   
#   Currently, JSOp::Retsub takes two arguments, and uses one to decide whether to throw the second or use it as a resume index. This patch splits out the exception-handling logic into explicit bytecode, leaving JSOp::Retsub as a simple indirect jump.
#   
#   To make the generated bytecode work out nicely, I've swapped the order of the arguments: now `throwing` is on top, and the exception/resumeindex argument is underneath it.
# ==============================================================================

diff -r 1cb0b352b024 -r eeaeb40f8b86 js/src/frontend/TryEmitter.cpp
--- a/js/src/frontend/TryEmitter.cpp	Tue Mar 08 21:10:22 2022 +0000
+++ b/js/src/frontend/TryEmitter.cpp	Tue Mar 08 21:10:22 2022 +0000
@@ -9,6 +9,7 @@
 #include "mozilla/Assertions.h"  // MOZ_ASSERT
 
 #include "frontend/BytecodeEmitter.h"  // BytecodeEmitter
+#include "frontend/IfEmitter.h"        // BytecodeEmitter
 #include "frontend/SharedContext.h"    // StatementKind
 #include "vm/Opcodes.h"                // JSOp
 
@@ -222,10 +223,27 @@
     }
   }
 
+  InternalIfEmitter ifThrowing(bce_);
+  if (!ifThrowing.emitThenElse()) {
+    return false;
+  }
+
+  if (!bce_->emit1(JSOp::Throw)) {
+    return false;
+  }
+
+  if (!ifThrowing.emitElse()) {
+    return false;
+  }
+
   if (!bce_->emit1(JSOp::Retsub)) {
     return false;
   }
 
+  if (!ifThrowing.emitEnd()) {
+    return false;
+  }
+
   bce_->hasTryFinally = true;
   return true;
 }